packages/client/tests/functional/_utils/relationMode/conditionalError.ts

```diff
@@ -0,0 +1,70 @@
+import { O } from 'ts-toolbelt'
+
+import { Providers } from '../providers'
+
+type RelationMode = 'prisma' | 'foreignKeys'
+
+type Target = {
+  provider: Providers
+  relationMode: RelationMode
+}
+
+// O.AtLeast<Record<Providers, string>>
+type ConditionalErrorSnapshotErrors = O.AtLeast<Record<Providers, string>> | string
+
+interface With<Supplied> {
+  with<T extends Omit<Target, keyof Supplied>, K extends keyof T>(
+    key: K,
+    value: T[K],
+  ): keyof Omit<Omit<Target, keyof Supplied>, K> extends never ? ConditionalErrorSnapshot : With<Supplied & Pick<T, K>>
+}
+
+interface ConditionalErrorSnapshot {
+  // snapshot(errors: O.AtLeast<Record<RelationMode, O.AtLeast<Record<Providers, string>>>>): string
+  snapshot(errors: O.AtLeast<Record<RelationMode, ConditionalErrorSnapshotErrors>>): string
+}
+
+class ConditionalErrorBuilder<Supplied> implements With<Supplied>, ConditionalErrorSnapshot {
+  constructor(private target: Partial<Target>) {}
+
+  with<T extends Omit<Target, keyof Supplied>, K extends keyof T>(key: K, value: T[K]) {
+    const target: Partial<Target> = { ...this.target, [key]: value }
+    return new ConditionalErrorBuilder<Supplied & Pick<T, K>>(target)
+  }
+
+  snapshot(errors: O.AtLeast<Record<RelationMode, ConditionalErrorSnapshotErrors>>) {
+    const { provider, relationMode } = this.target as Target
+    const errorBase = errors[relationMode]
+
+    if (typeof errorBase === 'string') {
+      return errorBase
+    }
+
+    if (errorBase === undefined) {
+      return `TODO: add error for relationMode=${relationMode}`
+    }
+
+    return errorBase[provider] || `TODO: add error for provider=${provider}`
```

Comment:
?

Reply:
If we do not have a snapshot of the error message in the test, we output `TODO: add error for relationMode=${relationMode}`

So it's obvious what needs to be done